src/util/metadata/vo_bit/mod.rs

```diff
@@ -160,3 +158,82 @@ pub fn bcopy_vo_bit_from_mark_bit<VM: VMBinding>(start: Address, size: usize) {
     let side_mark_bit_spec = mark_bit_spec.extract_side_spec();
     VO_BIT_SIDE_METADATA_SPEC.bcopy_metadata_contiguous(start, size, side_mark_bit_spec);
 }
+
+use crate::util::constants::{LOG_BITS_IN_BYTE, LOG_BYTES_IN_ADDRESS};
+
+/// How many data memory bytes does 1 word in the VO bit side metadata represents?
+pub const VO_BIT_WORD_TO_REGION: usize = 1
+    << (VO_BIT_SIDE_METADATA_SPEC.log_bytes_in_region
+        + LOG_BITS_IN_BYTE as usize
+        + LOG_BYTES_IN_ADDRESS as usize
+        - VO_BIT_SIDE_METADATA_SPEC.log_num_of_bits);
+
+/// Bulk check if a VO bit word. Return true if there is any bit set in the word.
+pub fn get_raw_vo_bit_word(addr: Address) -> usize {
+    unsafe { VO_BIT_SIDE_METADATA_SPEC.load_raw_word(addr) }
+}
+
+/// Find the base reference to the object from a potential internal pointer.
+pub fn find_object_from_internal_pointer<VM: VMBinding>(
+    start: Address,
+    search_limit_bytes: usize,
+) -> Option<ObjectReference> {
+    if !start.is_mapped() {
+        return None;
+    }
+
+    if let Some(vo_addr) = unsafe {
+        VO_BIT_SIDE_METADATA_SPEC.find_prev_non_zero_value::<u8>(start, search_limit_bytes)
+    } {
+        is_internal_ptr_from_vo_bit::<VM>(vo_addr, start)
+    } else {
+        None
+    }
+}
+
+/// Turning a potential object reference into its in-object address (the ref_to_address address) where the metadata is set for.
+fn get_in_object_address_for_potential_object<VM: VMBinding>(potential_obj: Address) -> Address {
+    // This is hacky: the address could be unaligned, and we will see an assertion failure if we
+    // create the 'object reference' in the normal way.
+    let obj_ref: ObjectReference = unsafe { std::mem::transmute(potential_obj) };
+    obj_ref.to_address::<VM>()
```

Comment:
We may use the constant `IN_OBJECT_ADDRESS_OFFSET` here.

Reply:
Yes. I will make changes once the other PR is merged.